test/node/src/join.test.ts

```diff
@@ -202,6 +202,70 @@ for (const dialect of DIALECTS) {
         ])
       })
 
+      it('should inner join a table using a complex `on` expression', async () => {
+        const query = ctx.db
+          .selectFrom('person')
+          .innerJoin('pet', (join) =>
+            join
+              .onRef('pet.owner_id', '=', 'person.id')
+              .on('pet.name', 'in', ['Catto', 'Doggo', 'Hammo'])
+              .on(({ or, cmp, ref, selectFrom }) =>
+                or([
+                  cmp('pet.species', '=', 'cat'),
+                  cmp('species', '=', 'dog'),
+                  cmp(
+                    ref('species'),
+                    '=',
+                    selectFrom('pet')
+                      .select(sql<'hamster'>`'hamster'`.as('hamster'))
```

Comment:
I wonder if we could make `sq.literal` infer the literal type so all a consumer would need to write is:

```ts
sql.literal('hamster').as('hamster')
```

And it would infer the return type as `'hamster'`.

Reply:
Yep, same with `eb.val`. Let's investigate this in another PR?

Reply:
Agreed!